packages/hardhat/test/e2e/crab-Migration.ts

```diff
@@ -0,0 +1,445 @@
+            const userEthBalanceAfter = await provider.getBalance(d4.address)
+
+            expect(crabV1SharesAfter.eq(0)).to.be.true
+            expect(crabV1SharesInMigration.eq(0)).to.be.true
+            expect(crabV2SharesAfter.gte(crabV2SharesBefore)).to.be.true // There will lil more crab because of the flash deposit
+            expect(userEthBalanceAfter.gt(userEthBalance)).to.be.true // Unused ETH from flash deposit will be returned
+            expect(crabV2ShraesInMigration).to.be.equal('0')
+        })
+
+        it("Should migrate d5 when CR1 < CR2", async () => {
+            await initialize(d5.address)
+            await decreaseCR1()
+            const [isFlashMigrate, ethNeededForV2, v1oSqthToPay, ethToGetFromV1] = await crabMigration.flashMigrationDetails(d5.address)
+
+
+            expect(wdiv(collatV1, shortV1).gt(wdiv(collatV2, shortV2))).to.be.false
+            expect(isFlashMigrate).to.be.false
+            expect(ethToGetFromV1.gt(ethNeededForV2)).to.be.false
+
+            const oSqthPrice = await oracle.getTwap(wethOsqthPoolAddress, squeethAddress, wethAddress, 1, false)
+
+            const numerator = ethToGetFromV1.sub(wmul(v1oSqthToPay, oSqthPrice).mul(101).div(100)) // 1% slippage
+            const denominator = one.sub(wdiv(oSqthPrice.mul(101).div(100), wdiv(collatV2, shortV2)))
+            const ethToFlashLoan = wdiv(numerator, denominator)
+            const expectedSqthFromV2 = wdiv(wmul(ethToFlashLoan, shortV2), collatV2)
+            const v1SqthToFlashWithdraw = v1oSqthToPay.sub(expectedSqthFromV2)
+            const maxEthToPay = wmul(v1SqthToFlashWithdraw, oSqthPrice).mul(101).div(100) // 1% slippage        
+
+            await crabStrategyV1.connect(d5).approve(crabMigration.address, ethers.constants.MaxUint256);
+            await expect(crabMigration.connect(d5).flashMigrateAndWithdrawFromV1toV2(one, 0, 0)).to.be.revertedWith("M8");
+            await expect(crabMigration.connect(d5).flashMigrateAndWithdrawFromV1toV2(one, one, 0)).to.be.revertedWith("M8");
+            await expect(crabMigration.connect(d5).flashMigrateAndWithdrawFromV1toV2(one, 0, one)).to.be.revertedWith("M8");
+            await expect(crabMigration.connect(d5).flashMigrateFromV1toV2(0)).to.be.revertedWith("M11");
+            // Don't flash deposit and return back ETH. For the sake of simplicity. Flash deposit is already tested in previous tests
+            await crabMigration.connect(d5).flashMigrateAndWithdrawFromV1toV2(0, ethToFlashLoan, maxEthToPay)
+
+            const crabV1SharesAfter = await crabStrategyV1.balanceOf(d5.address)
+            const crabV1SharesInMigration = await crabStrategyV1.balanceOf(crabMigration.address)
+            const crabV2ShraesInMigration = await crabStrategyV2.balanceOf(crabMigration.address)
+
+            expect(crabV1SharesAfter.eq(0)).to.be.true
+            expect(crabV1SharesInMigration.eq(0)).to.be.true
+            expect(crabV2ShraesInMigration).to.be.equal('0')
```

Comment:
Think here we need to check more things, like the above test.

User crab balances, eth balances, etc.